dataplane/linux/endpoint_mgr.go

```diff
@@ -489,111 +496,161 @@ func (m *endpointManager) resolveWorkloadEndpoints() {
 		m.needToCheckDispatchChains = true
 	}
 
-	// Update any dirty endpoints.
-	for id, workload := range m.pendingWlEpUpdates {
-		logCxt := log.WithField("id", id)
-		oldWorkload := m.activeWlEndpoints[id]
-		if workload != nil {
-			logCxt.Info("Updating per-endpoint chains.")
-			if oldWorkload != nil && oldWorkload.Name != workload.Name {
-				logCxt.Debug("Interface name changed, cleaning up old state")
-				m.epMarkMapper.ReleaseEndpointMark(oldWorkload.Name)
-				m.filterTable.RemoveChains(m.activeWlIDToChains[id])
-				m.routeTable.SetRoutes(oldWorkload.Name, nil)
-				m.wlIfaceNamesToReconfigure.Discard(oldWorkload.Name)
-				delete(m.activeWlIfaceNameToID, oldWorkload.Name)
-			}
-			var ingressPolicyNames, egressPolicyNames []string
-			if len(workload.Tiers) > 0 {
-				ingressPolicyNames = workload.Tiers[0].IngressPolicies
-				egressPolicyNames = workload.Tiers[0].EgressPolicies
-			}
-			adminUp := workload.State == "active"
-			chains := m.ruleRenderer.WorkloadEndpointToIptablesChains(
-				workload.Name,
-				m.epMarkMapper,
-				adminUp,
-				ingressPolicyNames,
-				egressPolicyNames,
-				workload.ProfileIds,
-			)
-			m.filterTable.UpdateChains(chains)
-			m.activeWlIDToChains[id] = chains
-
-			// Collect the IP prefixes that we want to route locally to this endpoint:
-			logCxt.Info("Updating endpoint routes.")
-			var (
-				ipStrings  []string
-				natInfos   []*proto.NatInfo
-				addrSuffix string
-			)
-			if m.ipVersion == 4 {
-				ipStrings = workload.Ipv4Nets
-				natInfos = workload.Ipv4Nat
-				addrSuffix = "/32"
-			} else {
-				ipStrings = workload.Ipv6Nets
-				natInfos = workload.Ipv6Nat
-				addrSuffix = "/128"
-			}
-			if len(natInfos) != 0 {
-				old := ipStrings
-				ipStrings = make([]string, len(old)+len(natInfos))
-				copy(ipStrings, old)
-				for ii, natInfo := range natInfos {
-					ipStrings[len(old)+ii] = natInfo.ExtIp + addrSuffix
+	removeActiveWorkload := func(logCxt *log.Entry, oldWorkload *proto.WorkloadEndpoint, id proto.WorkloadEndpointID) {
+		m.callbacks.InvokeRemoveWorkload(oldWorkload)
+		m.filterTable.RemoveChains(m.activeWlIDToChains[id])
+		delete(m.activeWlIDToChains, id)
+		if oldWorkload != nil {
+			m.epMarkMapper.ReleaseEndpointMark(oldWorkload.Name)
+			// Remove any routes from the routing table.  The RouteTable will remove any
+			// conntrack entries as a side-effect.
```

Comment:
If the interface is being recreated (as we thought was happening in OpenStack), is there any chance that this conntrack side-effect will cause problems?

Reply:
Well you are Mr Conntrack :smile: , but

- I guess it would be observable if someone was explicitly monitoring or depending on conntrack entries.

- Otherwise I would guess no, and that the worst case is going through iptables and NAT again?  (But if you go through NAT again, can you get a different result?)

- I guess we need to understand the customer scenario better, to work out if this matters.

Reply:
If you remove the conntrack entry that usually breaks any in-progress connections.  I think subsequent packets will hit the "conntrack invalid" drop rule so unfortunately it doesn't recover.

Probably OK if this really is a transient misconfiguration during a VM reboot or something (since the old connections will be gone anyway).  However, if Openstack just cycles the WEP for no reason, we might break some connections here.

Reply:
We've agreed for now to assume that it's OK to risk impacting existing connections - on the basis that the scenario is probably one of guest OS restart.